File: wasm/src/dbc/quotes.zig
//! Quoted-string parsing for DBC records.
//!
//! DBC uses quoted fields for units and value-description labels. This module
//! consumes one quoted field and leaves the caller's cursor at the next byte.

const std = @import("std");

/// Parses one quoted string, including escaped bytes, and advances `cursor`.
pub fn parseQuoted(allocator: std.mem.Allocator, cursor: *[]const u8) ![]const u8 {
    if (!std.mem.startsWith(u8, cursor.*, "\"")) return error.InvalidQuotedString;
    cursor.* = cursor.*[1..];

    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(allocator);

    while (cursor.*.len > 0) {
        const byte = cursor.*[0];
        cursor.* = cursor.*[1..];

        if (byte == '"') {
            return out.toOwnedSlice(allocator);
        }

        if (byte == '\\') {
            if (cursor.*.len == 0) return error.InvalidQuotedString;
            try out.append(allocator, cursor.*[0]);
            cursor.* = cursor.*[1..];
            continue;
        }

        try out.append(allocator, byte);
    }

    return error.InvalidQuotedString;
}

test "parse quoted string" {
    const allocator = std.testing.allocator;
    var cursor: []const u8 = "\"km/h\" Receiver";

    const parsed = try parseQuoted(allocator, &cursor);
    defer allocator.free(parsed);

    try std.testing.expectEqualStrings("km/h", parsed);
    try std.testing.expectEqualStrings(" Receiver", cursor);
}

test "parse escaped quoted string" {
    const allocator = std.testing.allocator;
    var cursor: []const u8 = "\"State \\\"On\\\" \\\\ A\" tail";

    const parsed = try parseQuoted(allocator, &cursor);
    defer allocator.free(parsed);

    try std.testing.expectEqualStrings("State \"On\" \\ A", parsed);
    try std.testing.expectEqualStrings(" tail", cursor);
}

File: wasm/src/dbc/quotes.rs
use super::DbcError;

/// Parses one quoted DBC field and returns it with the unconsumed input.
///
/// DBC escaping is deliberately byte-oriented: a backslash quotes the next
/// byte, matching the existing parser's behavior for units and value labels.
pub(crate) fn parse_quoted(cursor: &str) -> Result<(String, &str), DbcError> {
    let bytes = cursor.as_bytes();
    if bytes.first() != Some(&b'"') {
        return Err(DbcError::InvalidQuotedString);
    }

    let mut output = Vec::new();
    let mut index = 1;
    while index < bytes.len() {
        match bytes[index] {
            b'"' => {
                let parsed = String::from_utf8(output)
                    .expect("removing ASCII DBC escape bytes preserves valid UTF-8");
                return Ok((parsed, &cursor[index + 1..]));
            }
            b'\\' => {
                index += 1;
                let Some(&escaped) = bytes.get(index) else {
                    return Err(DbcError::InvalidQuotedString);
                };
                output.push(escaped);
                index += 1;
            }
            byte => {
                output.push(byte);
                index += 1;
            }
        }
    }

    Err(DbcError::InvalidQuotedString)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_quoted_string() {
        let (parsed, cursor) = parse_quoted("\"km/h\" Receiver").unwrap();

        assert_eq!(parsed, "km/h");
        assert_eq!(cursor, " Receiver");
    }

    #[test]
    fn parses_escaped_quoted_string() {
        let (parsed, cursor) = parse_quoted("\"State \\\"On\\\" \\\\ A\" tail").unwrap();

        assert_eq!(parsed, "State \"On\" \\ A");
        assert_eq!(cursor, " tail");
    }

    #[test]
    fn rejects_unterminated_quoted_string() {
        assert!(matches!(
            parse_quoted("\"unterminated"),
            Err(DbcError::InvalidQuotedString)
        ));
    }
}
